discovery-provider/src/tasks/entity_manager/user.py

```diff
@@ -0,0 +1,132 @@
+        raise Exception("Invalid User Transaction, wrong entity type")
+
+    if params.action == Action.CREATE:
+        if user_id in params.existing_records[EntityType.USER]:
+            raise Exception("Invalid User Transaction, user does not exist")
+        if user_id < USER_ID_OFFSET:
+            raise Exception("Invalid User Transaction, user id offset incorrect")
+    else:
+        # update / delete specific validations
+        if user_id not in params.existing_records[EntityType.USER]:
+            raise Exception("Invalid User Transaction, user does not exist")
+        wallet = params.existing_records[EntityType.USER][user_id].wallet
+        if wallet and wallet.lower() != params.signer.lower():
+            raise Exception(
+                "Invalid User Transaction, user wallet signer does not match"
+            )
+
+
+def update_user_record(params: ManageEntityParameters, user: User, metadata: Dict):
+    update_user_metadata(
+        params.session, params.redis, user, metadata, params.web3, params.challenge_bus
+    )
+    user.metadata_multihash = params.metadata_cid
+    user = update_legacy_user_images(user)
+    user = validate_user_record(user)
+    return user
+
+
+def create_user(params: ManageEntityParameters):
+    validate_user_tx(params)
+
+    user_id = params.user_id
+
+    user_record = User(
+        user_id=user_id,
+        wallet=params.signer.lower(),
+        txhash=params.txhash,
+        blockhash=params.event_blockhash,
+        blocknumber=params.block_number,
+        created_at=params.block_datetime,
+        updated_at=params.block_datetime,
+        is_current=False,
+    )
+
+    sp_ids = parse_sp_ids(params.metadata_cid)
+
+    # Update the user's new replica set in the model and save!
+    user_record.primary_id = sp_ids[0]
+    user_record.secondary_ids = sp_ids[1:]
+
+    # Update cnode endpoint string reconstructed from sp ID
+    creator_node_endpoint_str = get_endpoint_string_from_sp_ids(
+        params.redis, sp_ids[0], sp_ids[1:]
+    )
+    user_record.creator_node_endpoint = creator_node_endpoint_str
+
+    user_record = validate_user_record(user_record)
+    params.add_user_record(user_id, user_record)
+    return user_record
+
+
+def update_user(params: ManageEntityParameters):
+    validate_user_tx(params)
+
+    user_metadata = params.ipfs_metadata[params.metadata_cid]
+    user_id = params.entity_id
+    existing_user = params.existing_records[EntityType.USER][user_id]
+    existing_user.is_current = False  # invalidate
+    if (
+        user_id in params.new_records[EntityType.USER]
+    ):  # override with last updated playlist is in this block
+        existing_user = params.new_records[EntityType.USER][user_id][-1]
+
+    user_record = copy_user_record(
+        existing_user,
+        params.block_number,
+        params.event_blockhash,
+        params.txhash,
+        params.block_datetime,
+    )
+
+    if not user_record.handle:
```

Comment:
This is confusing to me. My understanding is

If there's no handle on the existing record in *this* block, query the db to see if there ever was a handle. If there is, we skip over it because there's already a user and the one updating here isn't the right one?

Maybe we can add a comment

Reply:
Well the Create User flow is now
1.) Call the entity manager create user flow with the parameters: user_id, wallet, sp_ids
2.) This is indexed and discovery knows about it / can expose the user's replica set via an endpoint - this enables content nodes to verify the user uploading content are from their primary
3.) Upload metadata to primary
4.) Call entity manager update user with the CID metadata - this has the user's handle - the first appearance of it, so we have to verify that the handle is unique here, otherwise reject the transaction.